File: desktop/web/src/ipc/adaptation.ts
import { invoke } from '@tauri-apps/api/core';
import { plantDbStatus } from '../app/health/state';
import { plantDbUnavailableMessage } from './plant-db-errors';
import type { CompatibilityResult, ReplacementSuggestion } from '../generated/contracts';

export type { CompatibilityResult, ReplacementSuggestion };

/** Returns true when plant DB is not available — all queries should short-circuit. */
function isDegraded(): boolean {
  return plantDbStatus.value !== 'available';
}

/**
 * Check a batch of species against a target hardiness zone.
 * Returns compatibility results for each species found in the DB.
 */
export async function checkPlantCompatibility(
  canonicalNames: string[],
  targetHardiness: number,
  locale = 'en',
): Promise<CompatibilityResult[]> {
  if (canonicalNames.length === 0) return [];
  if (isDegraded()) throw new Error(plantDbUnavailableMessage(plantDbStatus.value));
  return invoke<CompatibilityResult[]>('check_plant_compatibility', {
    canonicalNames,
    targetHardiness,
    locale,
  });
}

/**
 * Suggest replacement species for an incompatible plant.
 * Finds species with similar characteristics but compatible hardiness.
 */
export async function suggestReplacements(
  canonicalName: string,
  targetHardiness: number,
  limit = 5,
  locale = 'en',
): Promise<ReplacementSuggestion[]> {
  if (isDegraded()) throw new Error(plantDbUnavailableMessage(plantDbStatus.value));
  return invoke<ReplacementSuggestion[]>('suggest_replacements', {
    canonicalName,
    targetHardiness,
    limit,
    locale,
  });
}

File: desktop/src/commands/adaptation.rs
pub use crate::contracts::adaptation::{CompatibilityResult, ReplacementSuggestion};

/// Check a batch of species against a target hardiness zone.
///
/// A plant is compatible when the target zone falls within
/// `[hardiness_zone_min, hardiness_zone_max]`. Zone diff is the
/// minimum distance to the compatible range (0 when inside).
#[tauri::command]
pub fn check_plant_compatibility(
    plant_db: tauri::State<'_, crate::db::PlantDb>,
    canonical_names: Vec<String>,
    target_hardiness: i32,
    locale: String,
) -> Result<Vec<CompatibilityResult>, String> {
    crate::services::adaptation::check_plant_compatibility(
        &plant_db,
        canonical_names,
        target_hardiness,
        locale,
    )
}

/// Suggest replacement species for one that is incompatible at a target zone.
///
/// Finds species with the same stratum (if known) and similar height (within
/// +/- 50%) that are compatible with the target hardiness zone.
#[tauri::command]
pub fn suggest_replacements(
    plant_db: tauri::State<'_, crate::db::PlantDb>,
    canonical_name: String,
    target_hardiness: i32,
    limit: u32,
    locale: String,
) -> Result<Vec<ReplacementSuggestion>, String> {
    crate::services::adaptation::suggest_replacements(
        &plant_db,
        canonical_name,
        target_hardiness,
        limit,
        locale,
    )
}
